### PR TITLE
Add $requestHeaders parameters to necessary methods

### PR DESCRIPTION
**Pull request by julienbourdeau**
See internal conversation for context 
https://algolia.slack.com/archives/C3742QTHT/p1499242553433222

### CODE REVIEW

**Comment by maxiloc** [comment]
👍 but, like you, not in favor of merging it, it makes very little sense to have it for all methods

**Comment by rayrutjes** [comment]
This makes a lot of sense in languages that share the client between different requests.
In PHP though, I don't think we need this.

**Comment by maxiloc** [comment]
We need to add indexing methods

**Comment by rayrutjes** [comment]
alright will do that.

**Comment by rayrutjes** [comment]
This PR doesn't feel right.
Did we think of adding the info to the ClientContext @maxiloc ?

**Comment by rayrutjes** [comment]
We already have a way to do this: 

```
$index = $client->initIndex('your_index_name');
$client->setExtraHeader('X-Algolia-User-ID', 'user123');
$res = $index->search('query string');
```

I don't think we need more for now.

**Comment by maxiloc** [comment]
Looks good to me 👍 . @julienbourdeau Maybe double check it as well

File: src/AlgoliaSearch/Index.php [hunk 1]
```diff
@@ -682,7 +682,7 @@ private function getAlgoliaFiltersArrayWithoutCurrentRefinement($filters, $needl
      *
      * @return mixed
      */
-    public function searchForFacetValues($facetName, $facetQuery, $query = array(), $requestHeaders)
+    public function searchForFacetValues($facetName, $facetQuery, $query = array(), $requestHeaders = array())
```
Comment (julienbourdeau) [inline src/AlgoliaSearch/Index.php]: Good catch, thanks!

**Comment by julienbourdeau** [comment]
Looks good to me to 🚀

**Comment by alcaeus** [comment]
For what it's worth, this is a BC break that causes the AlgoliaSearchBundle for Symfony to fail:
```
  Warning: Declaration of Algolia\AlgoliaSearchBundle\Debug\DebugClient::doRe  
  quest($context, $method, $host, $path, $params, $data, $connectTimeout, $re  
  adTimeout) should be compatible with AlgoliaSearch\Client::doRequest($conte  
  xt, $method, $host, $path, $params, $data, $connectTimeout, $readTimeout, $  
  requestHeaders = Array)                                                      
```

Adding a parameter to a public method in a non-final class (even with default value) is a BC break that should not be done in a minor release.

**Comment by rayrutjes** [comment]
arf yes @alcaeus we will act fast.

**Comment by alcaeus** [comment]
Wow. I'm **extremely** disappointed. Let me break it down, maybe you understand my frustration:

> All classes should be considered final

"considered final" is not a thing. The class is either `final` or it isn't. If it is, do as you please. If it isn't, respect BC - simple as that. If you want to have the classes final, deprecate extending them in 1.24.0 and make them `final` in 2.0.

> and should be inherited by composition.

I can't inherit by composition if you don't provide an interface. The `Client` class needs to be extended in order for functionality to be added. Once you provide an interface, I'll add my changes by composing the class. It doesn't change anything though, as changing the method signature in an interface is only allowed in major releases as well, so it would've been a BC break either way.

> but we have to take a decision between having users with outdated version because they locked their version to 1, or introducing this breaking change.

"We have to either get people to update our software or force them to update theirs by breaking it". This is exactly why Semantic Versioning exists: I want to update from 1.21.0 -> 1.23.0 with confidence and without being afraid of BC breaks. Guess what: I'll be on 1.22.0 for a while because I don't appreciate you breaking my code with minor releases. This is exactly how you end up with users with outdated versions.

If you want to add a new parameter without breaking BC, take a look at how Symfony solves it: https://github.com/symfony/symfony/blob/3.4/src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php#L121..L134
TL;DR:
- Add param in docs only
- Read param using `func_num_args` and `func_get_args`
- Throw deprecation notice if method was overridden in an inherited class
- Fix properly in next major version

That's how you make changes like this one, not by breaking any custom implementation.

> Could you update the signature of the method where needed in your code or the library you are using

No. For now, I've locked the client dependency to `<1.23.0` - that keeps the code working and doesn't require me to fix your mistakes on our end. Eventually, maybe.

Just in case you're interested in the code I need to update: the `DebugClient` mentioned in the error message above is part of a contribution made by our team to your Symfony bundle (see https://github.com/algolia/AlgoliaSearchBundle/pull/57). Maybe having it merged would've prevented this kind of BC break in a minor release, maybe it wouldn't have.

**Comment by rayrutjes** [comment]
@alcaeus I appreciate the candid feedback here. We will see if using `func_get_args` could make sense here. I'm afraid it will make the code we have really unreadable but worth considering in order to avoid the breaking changes.
